src/app-layer-ssl.c

```diff
@@ -250,54 +255,416 @@ static void SSLSetTxDetectFlags(void *vtx, uint8_t dir, uint64_t flags)
     input += 2;
 
-    input += cipher_suites_length;
+    if (!(HAS_SPACE(cipher_suites_length)))
+        goto invalid_length;
+
+    if ((ssl_state->current_flags & SSL_AL_FLAG_STATE_CLIENT_HELLO) &&
+            ssl_config.enable_ja3) {
+        int rc;
+
+        JA3Buffer *ja3_cipher_suites = Ja3BufferInit();
+        if (ja3_cipher_suites == NULL)
+            return -1;
+
+        uint16_t processed_len = 0;
+        while (processed_len < cipher_suites_length)
+        {
+            if (!(HAS_SPACE(2))) {
+                Ja3BufferFree(ja3_cipher_suites);
+                goto invalid_length;
+            }
+
+            uint16_t cipher_suite = *input << 8 | *(input + 1);
+            input += 2;
+
+            if (TLSDecodeValueIsGREASE(cipher_suite) != 1) {
+                rc = Ja3BufferAddValue(ja3_cipher_suites, cipher_suite);
+                if (rc != 0) {
+                    Ja3BufferFree(ja3_cipher_suites);
+                    return -1;
+                }
+            }
+
+            processed_len += 2;
+        }
+
+        rc = Ja3BufferAppendBuffer(ssl_state->ja3_str, ja3_cipher_suites);
+        if (rc != 0) {
+            return -1;
+        }
+
+    } else {
+        /* Skip cipher suites */
+        input += cipher_suites_length;
+    }
+
+    return (input - initial_input);
+
+invalid_length:
+    SCLogDebug("TLS handshake invalid length");
+    SSLSetEvent(ssl_state,
+                TLS_DECODER_EVENT_HANDSHAKE_INVALID_LENGTH);
+    return -1;
+}
+
+static inline int TLSDecodeHSHelloCompressionMethods(SSLState *ssl_state,
+                                                     uint8_t *input,
+                                                     const uint32_t input_len)
+{
+    uint8_t *initial_input = input;
 
     if (!(HAS_SPACE(1)))
         goto invalid_length;
 
-    /* skip compression methods */
-    uint8_t compression_methods_length = *(input++);
+    /* Skip compression methods */
+    uint8_t compression_methods_length = *input;
+    input += 1;
+
+    if (!(HAS_SPACE(compression_methods_length)))
+        goto invalid_length;
 
     input += compression_methods_length;
 
-    /* extensions are optional (RFC5246 section 7.4.1.2) */
+    return (input - initial_input);
+
+invalid_length:
+    SCLogDebug("TLS handshake invalid_length");
+    SSLSetEvent(ssl_state,
+                TLS_DECODER_EVENT_HANDSHAKE_INVALID_LENGTH);
+    return -1;
+}
+
+static inline int TLSDecodeHSHelloExtensionSni(SSLState *ssl_state,
+                                               uint8_t *input,
+                                               const uint32_t input_len)
+{
+    uint8_t *initial_input = input;
+
+    /* There must not be more than one extension of the same
+       type (RFC5246 section 7.4.1.4). */
+    if (ssl_state->curr_connp->sni) {
+        SCLogDebug("Multiple SNI extensions");
+        SSLSetEvent(ssl_state,
+                TLS_DECODER_EVENT_MULTIPLE_SNI_EXTENSIONS);
+        return -1;
```

Comment:
should this be fatal or can we set warning and ignore the extra sni?

Reply:
We could just warn and continue decoding the rest of the rest of the packet. Changing the code so it does that instead.